src/number_theory/function_field.lean

```diff
@@ -107,4 +109,123 @@ is_integral_closure.is_dedekind_domain Fq[X] (ratfunc Fq) F _
+lemma infty_valuation.map_mul' (x y : ratfunc Fq) :
+  infty_valuation_def Fq (x * y) = infty_valuation_def Fq x * infty_valuation_def Fq y :=
+begin
+  rw [infty_valuation_def, infty_valuation_def, infty_valuation_def],
+  by_cases hx : x = 0,
+  { rw [hx, zero_mul, if_pos (eq.refl _), zero_mul] },
+  { by_cases hy : y = 0,
+    { rw [hy, mul_zero, if_pos (eq.refl _), mul_zero] },
+    { rw [if_neg hx, if_neg hy, if_neg (mul_ne_zero hx hy), ← with_zero.coe_mul,
+        with_zero.coe_inj, ← of_add_add],
+      apply congr_arg,
+      rw [add_sub, sub_add, sub_sub_assoc_swap, sub_sub, sub_eq_sub_iff_add_eq_add],
+      norm_cast,
+      rw [← polynomial.nat_degree_mul x.denom_ne_zero y.denom_ne_zero,
+        ← polynomial.nat_degree_mul (ratfunc.num_ne_zero (mul_ne_zero hx hy))
+          (mul_ne_zero x.denom_ne_zero y.denom_ne_zero),
+        ← polynomial.nat_degree_mul (ratfunc.num_ne_zero hx) (ratfunc.num_ne_zero hy),
+        ← polynomial.nat_degree_mul (mul_ne_zero (ratfunc.num_ne_zero hx) (ratfunc.num_ne_zero hy))
+          (x * y).denom_ne_zero, ratfunc.num_denom_mul],}}
+end
+
+omit dec
+
+variable {Fq}
+/-- Equivalent fractions have the same valuation. -/
+lemma infty_valuation_well_defined {r₁ r₂ s₁ s₂ : polynomial Fq} (hr₁ : r₁ ≠ 0) (hs₁ : s₁ ≠ 0)
+  (hr₂ : r₂ ≠ 0) (hs₂ : s₂ ≠ 0) (h_eq : r₁*s₂ = r₂*s₁) :
+  (r₁.nat_degree : ℤ) - s₁.nat_degree = (r₂.nat_degree : ℤ) - s₂.nat_degree :=
+begin
+  rw sub_eq_sub_iff_add_eq_add,
+  norm_cast,
+  rw [← polynomial.nat_degree_mul hr₁ hs₂, ← polynomial.nat_degree_mul hr₂ hs₁, h_eq],
+end
+
+lemma ratfunc.num_add_ne_zero {x y : ratfunc Fq} (hxy : x + y ≠ 0) :
+  x.num * y.denom + x.denom * y.num ≠ 0 :=
+begin
+  intro h_zero,
+  have h := ratfunc.num_denom_add x y,
+  rw [h_zero, zero_mul] at h,
+  exact (mul_ne_zero (ratfunc.num_ne_zero hxy) (mul_ne_zero x.denom_ne_zero y.denom_ne_zero)) h,
+end
+
+lemma infty_valuation_add_rw {x y : ratfunc Fq} (hxy : x + y ≠ 0) :
+  ((x + y).num.nat_degree : ℤ) - ((x + y).denom.nat_degree)  =
+  ((x.num) * y.denom + (x.denom) * y.num).nat_degree - ((x.denom) * y.denom).nat_degree :=
+infty_valuation_well_defined (ratfunc.num_ne_zero hxy) ((x + y).denom_ne_zero)
+    (ratfunc.num_add_ne_zero hxy) (mul_ne_zero x.denom_ne_zero y.denom_ne_zero)
+    (ratfunc.num_denom_add x y)
+
+lemma infty_valuation_rw {x : ratfunc Fq} (hx : x ≠ 0) {s : polynomial Fq} (hs : s ≠ 0):
+  (x.num.nat_degree : ℤ) - (x.denom.nat_degree)  =
+  ((x.num)*s).nat_degree - (s*(x.denom)).nat_degree :=
+begin
+  apply infty_valuation_well_defined (ratfunc.num_ne_zero hx) x.denom_ne_zero
+    (mul_ne_zero (ratfunc.num_ne_zero hx) hs) (mul_ne_zero hs x.denom_ne_zero),
+  rw mul_assoc,
+end
+
+include dec
+
+variable (Fq)
+lemma infty_valuation.map_add_le_max' (x y : ratfunc Fq) :
+  infty_valuation_def Fq (x + y) ≤ max (infty_valuation_def Fq x) (infty_valuation_def Fq y) :=
+begin
+  by_cases hx : x = 0,
+    { rw [hx, zero_add],
+      conv_rhs {rw [infty_valuation_def, if_pos (eq.refl _)]},
```

Comment:
Mathlib style is to indent inside the `{` but not before them:
```lean
  by_cases hx : x = 0,
  { rw [hx, zero_add],
    conv_rhs {rw [infty_valuation_def, if_pos (eq.refl _)]},
```
same throughout this proof